tardis/visualization/tools/rpacket_visualization.py

```diff
@@ -0,0 +1,691 @@
+        )
+        self.fig.update_yaxes(
+            range=[-1.1 * v_shells[-1], 1.1 * v_shells[-1]],
+            title="Velocity (km/s)",
+            exponentformat="none",
+            color=self.theme_colors[theme]["color"],
+            linecolor=self.theme_colors[theme]["linecolor"],
+            gridcolor=self.theme_colors[theme]["gridcolor"],
+            zerolinecolor=self.theme_colors[theme]["zerolinecolor"],
+        )
+
+        # adding the shells and photosphere
+        for shell_no in range(len(self.sim.model.radius.value)):
+            if shell_no == 0:
+                # photosphere
+                self.fig.add_shape(
+                    type="circle",
+                    xref="x",
+                    yref="y",
+                    x0=-1 * v_shells[shell_no],
+                    y0=-1 * v_shells[shell_no],
+                    x1=v_shells[shell_no],
+                    y1=v_shells[shell_no],
+                    line_color=self.theme_colors[theme][
+                        "photosphere_line_color"
+                    ],
+                    fillcolor=self.theme_colors[theme]["photosphere_fillcolor"],
+                    opacity=1,
+                )
+            elif shell_no == (len(self.sim.model.radius.value) - 1):
+                # outermost shell
+                self.fig.add_shape(
+                    type="circle",
+                    xref="x",
+                    yref="y",
+                    x0=-1 * v_shells[shell_no],
+                    y0=-1 * v_shells[shell_no],
+                    x1=v_shells[shell_no],
+                    y1=v_shells[shell_no],
+                    line_color=self.theme_colors[theme]["shells_line_color"],
+                    opacity=1,
+                )
+            else:
+                # remaining shells
+                self.fig.add_shape(
+                    type="circle",
+                    xref="x",
+                    yref="y",
+                    x0=-1 * v_shells[shell_no],
+                    y0=-1 * v_shells[shell_no],
+                    x1=v_shells[shell_no],
+                    y1=v_shells[shell_no],
+                    line_color=self.theme_colors[theme]["shells_line_color"],
+                    opacity=0.2,
+                )
+
+        # Adding packet trajectory
+
+        for packet_no in range(len(rpacket_x)):
+            self.fig.add_trace(
+                go.Scatter(
+                    x=rpacket_x[packet_no],
+                    y=rpacket_y[packet_no],
+                    mode="markers+lines",
+                    name="Packet " + str(packet_no + 1),
+                    showlegend=False,
+                    hovertemplate="<b>X</b>: %{x}"
+                    + "<br><b>Y</b>: %{y}<br>"
+                    + "<b>Last Interaction: %{text}</b>",
+                    text=[
+                        self.interaction_from_num.get(
+                            rpacket_interactions[packet_no][step_no]
+                        )
```

Comment:
What difference will this make, like using list of dictionaries instead of dictionary variable directly. Also, imo this would make things look more complex then now. What do you think?